provisioning/baremetal_pod.go

```diff
@@ -798,13 +783,25 @@ func EnsureMetal3Deployment(info *ProvisioningInfo) (updated bool, err error) {
 		info.EventRecorder, metal3Deployment, expectedGeneration)
 	if err != nil {
 		err = fmt.Errorf("unable to apply Metal3 deployment: %w", err)
-		return
+		// Check if ApplyDeployment failed because the existing Pod had an outdated
+		// Pod Selector.
+		selector, get_err := getMetal3DeploymentSelector(info.Client.AppsV1(), info.Namespace)
+		if get_err != nil || equality.Semantic.DeepEqual(selector, metal3Deployment.Spec.Selector) {
+			return
+		}
+		// This is an older deployment with the incorrect Pod Selector.
+		// Delete deployment now and re-create in the next reconcile.
+		// The operator is watching deployments so the reconcile should be triggered when metal3 deployment
+		// is deleted.
+		if delete_err := DeleteMetal3Deployment(info); delete_err != nil {
+			err = fmt.Errorf("unable to delete Metal3 deployment with incorrect Pod Selector: %w", delete_err)
+			return
+		}
 	}
-
 	if updated {
 		resourcemerge.SetDeploymentGeneration(&info.ProvConfig.Status.Generations, deployment)
```

Comment:
This code is now reachable in the event of an error due to a non-matching selector, so in that case we will set the expected generation back to 0. I've convinced myself that this is OK, since we are deleting the Deployment already. If ApplyDeployment finds the old Deployment gone, it will create a new one without looking at the generation; and if it finds the old one in the cache the selector won't match and it will try (unsuccessfully) to update again, also ignoring the generation.